### PR TITLE
impl(GCS+gRPC): a class for read responses

### PR DESCRIPTION
**Pull request by coryan**
Part of the work for #12116

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12677)

### CODE REVIEW

**Review by scotthart — APPROVED** [review]
> Reviewed 8 of 9 files at r1.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12677)* status: 8 of 9 files reviewed, 1 unresolved discussion (waiting on @coryan)

___
*[`google/cloud/storage/internal/async/read_payload_impl.cc` line 21 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12677#-NejRa9pAvJRflBQnVeT:-NejRa9q0YHuuzbrXvh-:bvpfo7c) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/bcfe09339687baec0f6ca002487a6c4676518fb4/google/cloud/storage/internal/async/read_payload_impl.cc#L21)):*
> ```cpp
> namespace storage_internal {
> GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
> ```

I'm assuming a future PR will add code here?

**Review by coryan — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12677)* status: 6 of 9 files reviewed, all discussions resolved (waiting on @scotthart)

___
*[`google/cloud/storage/internal/async/read_payload_impl.cc` line 21 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12677#-NejRa9pAvJRflBQnVeT:-Nej_ZlJAyV2dq5tedP2:bszny74) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/bcfe09339687baec0f6ca002487a6c4676518fb4/google/cloud/storage/internal/async/read_payload_impl.cc#L21)):*
<details><summary><i>Previously, scotthart (Scott Hart) wrote…</i></summary><blockquote>

I'm assuming a future PR will add code here?
</blockquote></details>

Maybe, but I should add the file at that point.

File: google/cloud/storage/internal/async/read_payload_impl.h [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+#include "google/cloud/storage/async_object_responses.h"
+#include "google/cloud/version.h"
+#include "absl/strings/cord.h"
+
+namespace google {
+namespace cloud {
+namespace storage_internal {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+
+struct ReadPayloadImpl {
```
Comment (devbww) [inline google/cloud/storage/internal/async/read_payload_impl.h:27]: It's probably a good idea to include the `_fwd.h` file to help validate its declaration(s).

File: google/cloud/storage/async_object_responses.h [hunk 2]
```diff
@@ -69,6 +73,98 @@ struct AsyncReadObjectRangeResponse {
...
+  /// The payload contents. These buffers are invalidated if this object is
+  /// modified.
+  std::vector<absl::string_view> contents() const {
+    return {impl_.chunk_begin(), impl_.chunk_end()};
+  }
+
+  /// The object metadata.
+  absl::optional<storage::ObjectMetadata> metadata() const { return metadata_; }
+
+  /// The starting offset the the current message.
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/storage/async_object_responses.h]: Done.

File: google/cloud/storage/internal/async/read_payload_impl.h (hunk 1 above)
Comment (coryan) [reply google/cloud/storage/internal/async/read_payload_impl.h:28]: Done.

**Pull request merged**